[우] 로또 1단계 - 자동 미션 제출합니다.

hwookim opened the pull request:
제 저장소에서 브랜치를 잘못 설정해 다시 보냅니다.

YerinCho committed:
[Lotto] docs : 구현 목록 작성

YerinCho committed:
[Lotto] feat : LottoFactory

 로또 번호 생성 기능 추가
 무작위로 6개 숫자 선택해 로또 생성

YerinCho committed:
[Lotto] refactor : LottoFactory

 상수화, 불필요한 코드 삭제

YerinCho committed:
[Lotto] docs : 구현 목록 수정

YerinCho committed:
[Lotto] feat : PurchaseLotto

 로또 구매 갯수 계산

YerinCho committed:
[Lotto] refactor : PurchaseLotto

 상수화

YerinCho committed:
[Lotto] docs : 구현목록 수정

YerinCho committed:
[Lotto] refactor : LottoFactoryTest

 불필요한 코드 제거

YerinCho committed:
[Lotto] feat : Lotto, 중복 검사 구현

YerinCho committed:
[Lotto] feat : InputView

YerinCho committed:
[Lotto] docs : InputView

YerinCho committed:
[Lotto] feat : InputView 입력 예외처리

YerinCho committed:
[Lotto] feat : PurchaseLotto 입력 예외처리

 입력한 금액이 1000원 미만이면 예외처리

YerinCho committed:
[Lotto] docs : README 수정

 구입 금액 입력 예외처리
 보너스볼 입력 예외처리

YerinCho committed:
[Lotto] feat : Buyer 생성자 구현

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : Lotto.validateSize

 6자리인지 검사

YerinCho committed:
[Lotto] feat : SplitLottoNumbers

 , 기준으로 String 분리

YerinCho committed:
[Lotto] feat : Lotto.validateNumberScope

 로또 번호의 범위 유효성 검사 추가

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : WinningLotto

 보너스볼 범위 유효성 검사 추가
 Lotto.validateNumbersScope 분리

YerinCho committed:
[Lotto] feat : WinningLotto 보너스볼 중복검사

 생성자 수정
 Lotto.lottoNumbers 접근제어자 변경

YerinCho committed:
[Lotto] docs : README 수정

YerinCho committed:
[Lotto] feat : WinnigResult 로또 일치 갯수 계산

YerinCho committed:
[Lotto] feat : WinningRankResult 보너스볼 일치 여부 확인

 WinnigResult -> WinningRankResult

YerinCho committed:
[Lotto] feat : WinningRankResult, WinningValue

 각 매수 별 당첨된 등수를 확인하는 기능 추가
 Lotto.getLottoNumbers 추가
 WinningLotto.isBonusNumber 추가
 WinningRankResult.isBonusNumberContain, WinningRankResult.checkCorrectNumber 의 파라미터 변경

YerinCho committed:
[Lotto] feat : RewardRate

 수익률 계산 기능 추가
 WinningRankResult.getWinningValueResult 추가

YerinCho committed:
[Lotto] refactor : RewardRate

 for 문 stream 으로 변경

YerinCho committed:
[Lotto] refactor : WinningRankResult.checkRank 메소드 분리

YerinCho committed:
[Lotto] refactor : Lotto.getLottoNumbers 불변성 보장

YerinCho committed:
`[Lotto] refactor : PurchaseLotto -> MoneyManager 클래스명 변경

 생성자 추가, 멤버변수 money 추가

YerinCho committed:
[Lotto] refactor : Buyer 멤버변수 제거

YerinCho committed:
[Lotto] fix : InputView 입력 버퍼 초기화

YerinCho committed:
[Lotto] feat : Buyer.getLottos 추가

YerinCho committed:
[Lotto] feat : 초기 구매 화면 구현

 LottoMain 추가
 LottoManager 추가
 OutputView 추가

YerinCho committed:
[Lotto] feat : 당첨번호, 보너스번호 입력 구현

 LottoManager.setWinningLotto 추가

YerinCho committed:
[Lotto] feat : 당첨 결과 화면 구현

 LottoManager.analyzeLotto 추가
 WinningRankResult.winningValueResult 자료형 LinkedHashMap 으로 변경
 WinningValue 메시지 추가

YerinCho committed:
[Lotto] fix : 입력 버퍼 오류 수정

YerinCho committed:
[Lotto] feat : 수익률 계산 결과 출력 구현

 수익률 계산식 수정
 LottoManager.analyzeRewardRate 추가
 MoneyManager.getMoney 추가

YerinCho committed:
[Lotto] feat : 예외 메시지 추가

YerinCho committed:
[Lotto] refactor

 메소드, 변수명, 클래스명 변경
 메소드 분리
 접근 제한자 변경
 원시값 상수화
 메소드가 위치하는 클래스 변경
 불필요한 코드 정리
 README 수정

YerinCho committed:
[Lotto] feat : LottoNumber,ConvertInput 추가

 Integer 형식의 로또 넘버를 LottoNumber 로 포장
 그에 따른 ConvertInput 수정

YerinCho committed:
[Lotto] refactor : 패키지 분리

YerinCho committed:
[Lotto] feat : ConvertInputTest 공백 제거 검사

YerinCho committed:
[Lotto] feat : LottoNumber

 로또 숫자 범위 확인 테스트

hwookim committed:
Merge pull request #1 from YerinCho/master

로또

Hue9010 commented on src/main/java/lotto/domain/lottoTicket/LottoFactory.java:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    public static List<LottoNumber> createLottoNumbers() {
```
`LottoFactory`라는 이름과 달리 `LottoNumbers`를 생성하는 메서드만 존재하네요. Class의 이름을 현재 가지고 있는 책임에 맞게 변경하거나 Class 이름 그대로 Lotto를 만들어주는게 좋을 것 같네요. 😺

hwookim replied:
로또 번호 6자리를 생성하는 객체이므로 LottoNumberFactory로 변경하였습니다.

Hue9010 commented on src/main/java/lotto/domain/lottoTicket/LottoNumber.java:
```diff
@@ -0,0 +1,39 @@
+
+    static void validateNumberScope(int lottoNumber) {
+        if (lottoNumber > MAX_LOTTO_NUMBER || lottoNumber < MIN_LOTTO_NUMBER) {
+            throw new IllegalArgumentException(LOTTO_NUMBER_SCOPE_ERROR_MESSAGE);
```
교육 자료를 보면 커스텀 Exception을 학습한 걸로 보이는데 그게 맞다면 한번 RuntimeException을 상속받은 LottoNumberException을 만들어보세요!

Hue9010 replied:
마찬가지로 기본 Exception들을 커스텀 Excetpion으로 변경해보죠!

hwookim replied:
custom exception 클래스를 각각 만들어 예외 상황의 이름이 보여지도록 변경했습니다.

각 클래스의 위치를 어떻게 하는게 좋을까요?
패키지를 따로 분리하는 게 나을까요? 아니면 사용되는 클래스와 함께 두는게 나을까요?
일단 util에 넣어두긴 했다만... 아직 패키지 분리를 어떻게 해야 좀 더 나을지 잘 모르겠네요ㅠㅠ

많아질 경우를 대비해서 exception 패키지가 따로 있는게 나으려나요?
그렇다면 그건 domain 안에 있어야 할까요 밖에 있어야 할까요

Hue9010 replied:
지금과 같이 복잡하지 않고 단순한 패키지 구조라면 최상위 패키지인(정말 최상위 말고!) lotto에 exception 패키지를 만들어 위치시켜 둘 것 같습니다.

Hue9010 commented on src/main/java/lotto/domain/lottoTicket/WinningLotto.java:
```diff
@@ -0,0 +1,26 @@
+
+    public boolean isBonusNumber(Lotto lotto) {
+        return lotto.getLottoNumbers()
+                .contains(this.bonusNumber);
```
Lotto의 LottoNumbers를 getter로 가져오지 말고 lotto.contains를 통해 보너스 번호 존재 여부 확인은 Lotto한테 맡기면 어떨까요? 😺

hwookim replied:
Lotto 내에서 LottoNumber를 비교해 boolean을 반환하는 contains 메소드를 만들고 getter를 제거했습니다.
getter를 어떻게 줄일 수 있을지 고민하고 있었는데 좋은 방법이네요! 왜 이 생각을 못했지.....
감사합니다:)

Hue9010 commented on src/main/java/lotto/domain/result/LottoResult.java:
```diff
@@ -0,0 +1,49 @@
+public class LottoResult {
+    private static final int RATE = 100;
+    private static final int ZERO = 0;
+    private static final int ONE = 1;
```
매직 넘버로서 사용되고 있는 것도 아닌 0, 1 그대로의 의미를 가지고 있다면 숫자의 사용이 많다던지 일부 특수한 경우가 아니라면 굳이 ZERO, ONE으로 바꿀 필요가 있나 생각해볼 여지가 있겠네요.

hwookim replied:
모든 원시값을 상수화한다는 생각에 사로잡혀 조금 쓸데 없는 부분까지 적용해버린 것 같네요
다음부턴 필요한 상수화인지 한번 더 고민해보도록 하겠습니다 😄

Hue9010 replied:
숫자가 있으면 안된다가 아니라 매직넘버를 숫자그대로 사용하지 않는다고 생각하시면 좋답니다!

Hue9010 commented on src/main/java/lotto/domain/result/LottoResult.java:
```diff
@@ -0,0 +1,49 @@
+
+    public Map<WinningValue, Integer> getLottoResult() {
+        lottoResult.remove(WinningValue.ZERO);
+        return Collections.unmodifiableMap(lottoResult);
```
unmodifiableMap 👍 
`Collections.unmodifiableMap`를 어떤 목적으로 사용하셨는지 혹은 이와 관련된 부분에 대해서 궁금한 부분이 있는지 궁금합니다. 😺

hwookim replied:
Collection에 대하여 getter를 사용하게 되면 재할당은 불가할지 몰라도 put이나 remove 등의 내용 변경이 가능합니다.
이를 방지하고 객체 필드 변수의 불변성을 유지하기 위해서 unmodifiable 메소드를 사용했습니다.

이것보다는 getter를 사용하지 않고 객체 내에서 계산에 대한 결과값을 반환하도록 하는게 더 안전하게 데이터를 관리할 수 있다고 생각하지만.... 아직 익숙치 않은지 무심코 getter를 쓰게되네요 ㅠㅠ

Hue9010 replied:
사실 문의 드린 이유는 여기서 나올 수 있는 얘기가 많다보니 이 부분을 어느 목적으로 작성한지 알고 싶었네요 :)

제가 잘못 이해 한게 아니라면 `LottoResult`내의 `lottoResult`를 불변객체로 관리하고 싶었던게 맞을까요?

```java
private Map<WinningValue, Integer> lottoResult = new LinkedHashMap<>();
```
그렇다면 lootoResult에 final 키워드가 존재하지 않아 다른 Map으로 변경될 수 있는 가능성이 있습니다.

두번째로는 `LottoResult`내에서는 `lottoResult` 값이 바뀔 수 있는데 불변하다고 할 수 있을까요? 

문의 드렸던 목적이 여기서 나옵니다.
1. `LottoResult`는 자신의 상태인 `lottoResult`를 변경 할 수 있으나 외부에서 바꾸는건 막고 싶다.
2. `lottoResult`는 불변객체로 둔다.
외부에 `lottoResult`를 반환해줄 때 변경이 불가능한 Collection으로 내려줌으로 1번의 상황은 만족시킨 상태입니다. 그렇다면 의도한대로 구현하셨을 수 있지만 2번의 상황을 의도하신 거라면 정말 불변한지 생각해볼 여지가 있겠네요.

1번, 2번어느게 맞다가 아니라 의도한대로 구현되었냐가 핵심입니다.

그리고 getter가 필요한 상황이면 사용해야지요! 😸 불필요한 getter를 안만드는게 중요하답니다. 
\+ getter보다는 상태를 바꿀 수 있는 setter 사용에 더 주의를 요하는게 좋답니다!

(참고로 리뷰는 리뷰 요청 온 다음 진행 할게요 😋)

hwookim replied:
제가 불변성이라는 단어를 제가 잘못 사용한 것 같네요. 😢 
말씀해주신 1번, 2번 중에서는 1번만을 의도하고 작성한 코드입니다!

`LottoResult` 내에서는 각 등수별 횟수를 저장하기 위해 `lottoResult`를 변경할 필요가 있습니다.
다만 외부에서 `lottoResult`에 접근할 때는 이미 계산을 모두 끝마친 뒤기에 변경을 막으려는 의도였습니다.

아직 수정할 점이 많아서 조금 더 다듬고 리뷰 요청 보내도록 할게요!
감사합니다 😄

Hue9010 replied:
넵 그래서 말그대로 문의를 먼저 드렸던 겁니다 👍

Hue9010 commented on src/main/java/lotto/domain/lottoTicket/LottoFactory.java:
```diff
@@ -0,0 +1,28 @@
+
+    private static List<LottoNumber> lottoNumbers = new ArrayList<>();
+
+    public LottoFactory() {
```
LottoFactory를 정상적으로 사용하려면 new LottoFactory()후에나 사용가능하지 않던가요?

https://www.hanumoka.net/2017/09/13/java-20170913-java-static-initialization/
한번 이 자료 참고해서 new LottoFactory 부분 제거해보세요! 😄

Hue9010 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -0,0 +1,38 @@
+        System.out.println(BONUS_NUMBER_INPUT_MESSAGE);
+        try {
+            return Integer.parseInt(SCANNER.nextLine());
+        } catch (InputMismatchException e) {
```
InputMismatchException으로 선택하신 이유가 있을까요?
+ 혹시 테스트 중에 이 예외가 발생했다면 어떤 상황에서 나는 예외일까요?

hwookim replied:
처음에 nextInt를 통해 정수 값을 받도록 해서 InputMismatchException을 사용했었습니다.
입력버퍼 문제를 해결하기 위해 nextLine으로 변경했는데, 위의 inputMoney는 바꿔놓고 밑의 메소드는 변경하지 않았네요.

NumberFormatException으로 변경했습니다.

Hue9010 replied:
불필요해진 부분은 그때 안 없애면 까먹고 그대로 남거나 이걸 지워도 되는지 다시 생각해보는 시간으로 인해 생각보다 많은 낭비가 발생 할 수 있답니다.

Hue9010 commented on src/main/java/lotto/domain/lottoTicket/Lotto.java:
```diff
@@ -0,0 +1,52 @@
+                .count();
+    }
+
+    //출력을 위해 Integer로 변환해서 반환
```
`LottoNumber`의 toString을 오버라이드 하면 어떻게 될까요? 😺

hwookim replied:
`LottoNumber`와 `Lotto`에 toString 메소드를 오버라이드해 추가했습니다.

Hue9010 commented on src/main/java/lotto/controller/LottoManager.java:
```diff
@@ -0,0 +1,39 @@
+
+public class LottoManager {
+    public Money money;
+    public Buyer buyer;
```
getter를 안만들기 위해서 public으로 만들고 직접 접근하게 만든걸까요?
자신의 상태를 public으로 두고 외부에 공개해두면 자신도 모르게 값이 바뀔 수도 있답니다.

hwookim replied:
`money.calculateLottoTicketCount()`와 `buyer.getLottos()`에 접근할 수 있도록 메소드를 추가했습니다.

Hue9010 commented on src/main/java/lotto/domain/lottoTicket/Lotto.java:
```diff
@@ -0,0 +1,52 @@
+        this.lottoNumbers = numbers;
+    }
+
+    static void validateDuplication(List<LottoNumber> numbers) {
```
테스트를 위해서 static에 default 메서드로 만드신 것 같은데 생성시에 올바른 값인지 검증하게 되어 있으니까 validation 테스트는 생성자를 호출하는 방식으로 바꾸고, 그에 따라 접근 제어자랑 불필요해지는 static 제거해주세요! (다른 클래스들도요!)

테스트를 위한 변경도 없는게 좋답니다.

hwookim replied:
검증하는 부분을 모두 private void로 변경하고, 테스트코드에서는 생성자를 통해 테스트가 이뤄지도록 변경했습니다.

Hue9010 commented on src/test/java/lotto/domain/lottoTicket/LottoNumberTest.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Test
+    @DisplayName("로또 숫자 범위 확인")
+    void validateNumberScopeTest() {
```
경계값에 대한 테스트 잘해 주셨네요! 👍 
마찬가지로 성공하는 경우의 경계값인 1, 45도 테스트 해주는게 좋답니다. 덤으로 지금은 테스트 이름이 성공하는 케이스인지 실패인지 알 수가 없는데 두개를 구분해 주세요!

hwookim replied:
한가지 테스트 안에서 성공과 실패를 모두 검사하는 것으로 바꾸긴 했는데
테스트를 아예 두개로 분리하는게 나으려나요?

Hue9010 replied:
정말 아주 단순하다면 같이 할 법도 하지만 성공적으로(?) 실패하는 걸 확인하고자 분리하는 편입니다.
거기다 성공과 실패가 같이 있는 경우 메시지를 통해 알 수 있긴 해도 직관적으로 실패한 케이스의 이름만 봐도 알 수있는게 버그 수정에 용이해서 나눕니다.
+ Junit4 환경에서는 assertThatThrownBy를 사용하지 못해서 구분을 하는게 더 낫기도합니다.

Hue9010 reviewed (changes requested):
이런 리뷰를 맡은 휴입니다!
개인 사정으로 피드백이 늦었네요 😭 
군더더기 없이 깔끔히 구현해주셨는데 몇가지 피드백 남겨드렸습니다. 확인하시고 궁금하신 사항 생기면 코멘트나 DM으로라도 문의 주시면 됩니다! 😄

hwookim committed:
[Lotto] refactor : LottoFactory -> LottoNumberFactory

hwookim committed:
[Lotto] refactor : LottoFactory -> LottoNumberFactory

hwookim committed:
[Lotto] refactor : Custom Exception 추가

 InvalidLottoNumberException : LottoNumber의 범위 오류
 DuplicationLottoException : Lotto.lottoNumbers의 중복 오류
 InvalidSizeLottoException : Lotto.lottoNumbers의 사이즈 오류
 InputView.inputBonusNumber의 Exception 변경
  : InputMismatchException -> NumberFormatException

hwookim committed:
[Lotto] refactor : getter를 대신하는 contains 메소드 추가

 getter를 사용하지 않고, 객체에 질문하여 결과값을 도출할 수 있도록 변경

hwookim committed:
[Lotto] refactor : Custom Exception 추가

 InvalidBonusNumberException : 보너스 번호 중복 오류
 InvalidMoneyException : 1000원 미만 입력 오류

hwookim committed:
[Lotto] refactor : InvalidMoneyException 매개변수 추가

hwookim committed:
[Lotto] refactor : 무의미한 상수 제거

hwookim committed:
[Lotto] refactor : static을 통한 생성자 호출 제거

hwookim committed:
[Lotto] refactor : Custom Exception 추가에 따른 테스트 변경

hwookim committed:
[Lotto] refactor : 접근제어자 변경

 validation 메소드 접근 제어자 변경 default -> private
 생성자를 통해 테스트 진행하도록 테스트코드 변경

hwookim committed:
[Lotto] refactor : LottoResult.LottoResult final 추가

 재할당 방지

hwookim committed:
[Lotto] refactor : Lotto.getLotto -> toString

hwookim committed:
[Lotto] refactor : 불필요한 코드 제거

hwookim committed:
[Lotto] refactor : 가독성 향상을 위한 _ 사용

hwookim committed:
[Lotto] refactor : 필드변수 접근제어자 변경 public -> private

 main에서 접근하기 위해 getter 추가

Hue9010 commented on src/main/java/lotto/domain/lottoTicket/LottoNumberFactory.java:
```diff
@@ -0,0 +1,28 @@
+import java.util.Collections;
+import java.util.List;
+
+public class LottoNumberFactory {
```
구현하실 때 LottoNumberFactory의 책임에 대한 부분을 주석으로 작성하시고 작업하신 것 같은데 TODO, FIXME 를 활용할 수도 있을 것 같네요.(아주 잠시 사용한걸로 보여 이번에는 효율적이지 않았을 것 같지만 알아두면 좋을 것 같네요) 

다만 주석 내 내용도 "로또번호 6자리를 생성하는 객체로 구현"와 같이 TODO/FIXME에 해당하는 내용으로 적어주는게 좋을 것 같네요. 다만 무의미한 TODO를 늘려가면 의미있는 TODO가 묻혀버리니 남용은 자제하는게 좋답니다.

https://cheese10yun.github.io/intellij-todo/

Hue9010 commented on src/main/java/lotto/util/InvalidLottoNumberException.java:
```diff
@@ -0,0 +1,7 @@
+
+public class InvalidLottoNumberException extends RuntimeException {
+    public InvalidLottoNumberException(int lottoNumber) {
+        super(lottoNumber + " : 로또 번호는 1 ~ 45를 벗어날 수 없습니다.");
```
이정도는 자바에서 알아서 최적화 해 줄 것으로 기대할 수 있긴한데 String.format()을 사용해도 좋답니다

Hue9010 reviewed (approved):
잘 반영해주셨네요! 💯 
이번 단계는 머지하도록 할게요~
추가로 궁금하신 사항이 있으면 코멘트나 DM으로 문의주세요~!!
